lib/upgrade/etcd/conversionv1v3/rule.go

```diff
@@ -0,0 +1,266 @@
+		SrcNet:      ar.Source.Net,
+		SrcNets:     ar.Source.Nets,
+		SrcSelector: ar.Source.Selector,
+		SrcPorts:    ar.Source.Ports,
+		DstTag:      ar.Destination.Tag,
+		DstNet:      normalizeIPNet(ar.Destination.Net),
+		DstNets:     normalizeIPNets(ar.Destination.Nets),
+		DstSelector: ar.Destination.Selector,
+		DstPorts:    ar.Destination.Ports,
+
+		NotSrcTag:      ar.Source.NotTag,
+		NotSrcNet:      ar.Source.NotNet,
+		NotSrcNets:     ar.Source.NotNets,
+		NotSrcSelector: ar.Source.NotSelector,
+		NotSrcPorts:    ar.Source.NotPorts,
+		NotDstTag:      ar.Destination.NotTag,
+		NotDstNet:      normalizeIPNet(ar.Destination.NotNet),
+		NotDstNets:     normalizeIPNets(ar.Destination.NotNets),
+		NotDstSelector: ar.Destination.NotSelector,
+		NotDstPorts:    ar.Destination.NotPorts,
+	}
+}
+
+// normalizeIPNet converts an IPNet to a network by ensuring the IP address is correctly masked.
+func normalizeIPNet(n *net.IPNet) *net.IPNet {
+	if n == nil {
+		return nil
+	}
+	return n.Network()
+}
+
+// normalizeIPNets converts an []*IPNet to a slice of networks by ensuring the IP addresses
+// are correctly masked.
+func normalizeIPNets(nets []*net.IPNet) []*net.IPNet {
+	if nets == nil {
+		return nil
+	}
+	out := make([]*net.IPNet, len(nets))
+	for i, n := range nets {
+		out[i] = normalizeIPNet(n)
+	}
+	return out
+}
+
+// ruleBackendToAPI convert a Backend Rule structure to an API Rule structure.
+func ruleBackendToAPI(br model.Rule) apiv3.Rule {
+	var icmp, notICMP *apiv3.ICMPFields
+	if br.ICMPCode != nil || br.ICMPType != nil {
+		icmp = &apiv3.ICMPFields{
+			Code: br.ICMPCode,
+			Type: br.ICMPType,
+		}
+	}
+	if br.NotICMPCode != nil || br.NotICMPType != nil {
+		notICMP = &apiv3.ICMPFields{
+			Code: br.NotICMPCode,
+			Type: br.NotICMPType,
+		}
+	}
+
+	srcNets := br.AllSrcNets()
+	var srcNetsStr []string
+	for _, net := range srcNets {
+		srcNetsStr = append(srcNetsStr, net.String())
+	}
+
+	dstNets := br.AllDstNets()
+	var dstNetsStr []string
+	for _, net := range dstNets {
+		dstNetsStr = append(dstNetsStr, net.String())
+	}
+
+	notSrcNets := br.AllNotSrcNets()
+	var notSrcNetsStr []string
+	for _, net := range notSrcNets {
+		notSrcNetsStr = append(notSrcNetsStr, net.String())
+	}
+
+	notDstNets := br.AllNotDstNets()
+	var notDstNetsStr []string
+	for _, net := range notDstNets {
+		notDstNetsStr = append(notDstNetsStr, net.String())
+	}
+
+	// Tags are deprecated in v3.0+, so we convert Tags to selectors.
+	// For example:
+	// A rule that looks like this with Calico v1 API:
+	// apiv1.Rule{
+	//	Action: "allow",
+	//	Source: apiv1.EntityRule{
+	//		Tag:      "tag1",
+	//		Selector: "label1 == 'value1' || make == 'cake'",
+	//	},
+	//}
+	// That rule will be converted to the following for Calico v3 API:
+	// apiv3.Rule{
+	//	Action: "allow",
+	//	Source: apiv3.EntityRule{
+	//		Selector: "(label1 == 'value1' || make == 'cake') && tag1 == ''",
+	//	},
+	//}
+	srcSelector := br.SrcSelector
+	if br.SrcTag != "" {
```

Comment:
This isn't quite right - try writing a test for a rule that looks like this (no selector):

```go
apiv1.rule{
  Action: "allow",
  Source: apiv1.EntityRule{Tag: "tag1"}
}
```

I think you'll find the result to have this:

```
Selector: () && tag1 == ""
```

When it should be this:

```
Selector: tag1 == ""
```

This applies for all of the following cases (dstSelector, notSrcSelector, etc) as well.

Reply:
yup, you're totally right. I've fixed it